### DISCUSSION TITLE
How to properly handle rdkit's ArgumentError

### DISCUSSION THREAD

**Discussion by hogru** [rdkit/rdkit#5470] [Q&A] [answered]
I am playing around with improving my handling of invalid SMILES strings, more specifically do it without `m = Chem.MolFromSmiles(smiles)`.

This is a code snippet I currently use:

```py
from rdkit import Chem, rdBase

smiles = "Invalid SMILES"
rdBase.DisableLog("rdApp.error")
try:
    smiles = Chem.CanonSmiles(smiles)
except:  # noqa: E722
    pass  # Do something meaningful, omitted
finally:
    rdBase.EnableLog("rdApp.error")
```

What I don't like here is the generic except clause. For invalid SMILES, `rdkit` seems to throw an `ArgumentError`. So I tried this (changes denoted with "<--")

```py
from rdkit import Chem, rdBase
from ctypes import ArgumentError  # <--

smiles = "Invalid SMILES"
rdBase.DisableLog("rdApp.error")
try:
    smiles = Chem.CanonSmiles(smiles)
except ArgumentError as e:  # <--
    pass  # Do something meaningful, omitted
except Exception as e:  # <--
    raise RuntimeError("Unexpected rdkit exception") from e  # <--
finally:
  rdBase.EnableLog("rdApp.error")
```

But _this_ `ctypes.ArgumentError` is never raised. I also tried `argparse.ArgumentError` just to be sure.

Can you please tell me: what is the correct way to catch `rdkit`s `ArgumentError`s ?

**Comment by ptosco** [comment, marked as answer]
@hogru If you look at the stack trace, you will see that `CanonSmiles()` is a simple Python wrapper that calls first `MolFromSmiles()` and then `MolToSmiles()`:
```
ArgumentError                             Traceback (most recent call last)
/tmp/ipykernel_113176/1451563720.py in <module>
----> 1 Chem.CanonSmiles(smiles)

~/install/rdkit_github/master/lib/python3.8/site-packages/rdkit/Chem/__init__.py in CanonSmiles(smi, useChiral)
     55 def CanonSmiles(smi, useChiral=1):
     56   m = MolFromSmiles(smi)
---> 57   return MolToSmiles(m, useChiral)
     58 
     59 

ArgumentError: Python argument types in
    rdkit.Chem.rdmolfiles.MolToSmiles(NoneType, int)
did not match C++ signature:
    MolToSmiles(RDKit::ROMol mol, bool isomericSmiles=True, bool kekuleSmiles=False, int rootedAtAtom=-1, bool canonical=True, bool allBondsExplicit=False, bool allHsExplicit=False, bool doRandom=False)
    MolToSmiles(RDKit::ROMol mol, RDKit::SmilesWriteParams params)
```
As the initial call to `MolFromSmiles` returns `None` for your invalid SMILES, the subsequent call to `MolToSmiles` is passed `None` and hence `boost::python` throws a `Boost.Python.ArgumentError`.

The simplest way to catch the error condition is refactoring your code as:
```
smiles = "Invalid SMILES"
m = Chem.MolFromSmiles(smiles)
if m is None:
    pass  # Do something meaningful, omitted
else:
    smiles = Chem.MolToSmiles(m)
```

**Reply by hogru** [reply]
Thank you, @ptosco ! I was not aware of the wrapper (apparently)... ahem, changed, works :-)
